libs/core/type_support/include/hpx/type_support/uninitialized_relocate.hpp

```diff
@@ -0,0 +1,127 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/type_support/is_trivially_relocatable.hpp>
+#include <hpx/type_support/relocate_at.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+namespace hpx {
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+    using std::uninitialized_relocate;    // what would be the include for this?
+#else
+
+    namespace detail {
+
+        template <typename InIter, typename FwdIter>
+        struct choose_uninitialized_relocate_impl
+        {
+            using object_type =
+                std::decay_t<typename std::iterator_traits<InIter>::reference>;
+            using out_type =
+                std::decay_t<typename std::iterator_traits<FwdIter>::reference>;
+
+            using is_buffer_memcpyable =
+                std::bool_constant<std::is_same_v<object_type, out_type> &&
+                    hpx::is_trivially_relocatable_v<object_type> &&
+                    !std::is_volatile_v<object_type> &&
+                    !std::is_volatile_v<out_type> &&
+                    std::is_pointer_v<InIter> && std::is_pointer_v<FwdIter>>;
+
+            using type = std::integral_constant<int,
+                    is_buffer_memcpyable::value ? 0 : // single memcpy
+                    std::is_nothrow_constructible_v<
+                        out_type, std::add_rvalue_reference_t<object_type>
+                    >                                                 ? 1 : // loop without try-catch
+                                                                        2>; // loop with try-catch
+        };
+
+        template <typename InIter, typename FwdIter,
+            std::enable_if_t<choose_uninitialized_relocate_impl<InIter,
+                                 FwdIter>::type::value == 0,    // single memcpy
+                int> = 0>
+        FwdIter uninitialized_relocate_helper(
+            InIter first, InIter last, FwdIter dest)
+        {
+            auto n_objects = std::distance(first, last);
+
+            if (n_objects != 0)
+            {
+                std::byte* first_byte =
+                    reinterpret_cast<std::byte*>(std::addressof(*first));
+                std::byte* last_byte =
+                    reinterpret_cast<std::byte*>(std::addressof(*last));
+
+                auto n_bytes = std::distance(first_byte, last_byte);
+
+                memmove(dest, first, n_bytes);
```

Comment:
Thank you, hadn't thought of that.

So it makes sense to modify `relocate_at` to accept different input/output types?
```c++
template <typename SourceType, typename DestType>
T* relocate_at(SourceType* src, DestType* dst)
```

Reply:
I do that in libc++ internally: implement `relocate_at(T*, T*)` in terms of `__libcpp_relocate_at2(T*, U*)`. But I have never proposed that in P1144R0–R8, because `relocate_at(T*, U*)` can't ever take advantage of _trivial_ relocatability unless `T==U`. (I mean, in theory, a sufficiently smart implementation could expose `__is_trivially_relocatable_from(T, U)`; but nobody does that, and it's not at all clear what would be the syntax via which the user could "warrant" such a property themselves. See https://quuxplusone.github.io/blog/2018/07/03/trivially-constructible-from/ for more thoughts on the topic.)

Note that the "relocatable-from" relationship, if it existed, wouldn't be symmetrical:
```
    int *a;
    std::unique_ptr<int> b;
    std::construct_at(&b, std::move(a)); std::destroy_at(&a); // OK
    std::construct_at(&a, std::move(b)); std::destroy_at(&b); // Ill-formed

    static_assert(std::is_relocatable_from_v<std::unique_ptr<int>, int*>);
    static_assert(!std::is_relocatable_from_v<int*, std::unique_ptr<int>>);
```

...However, maybe you just meant "shouldn't it be possible to relocate _from_ const _into_ non-const?"—
```
    const int j = 42;
    alignas(int) char ibuf[sizeof(int)];
    int& i = *(int*)ibuf;
    std::relocate_at(&j, &i);  // P1144R8: Ill-formed, deduction failed: `const int` versus `int`
```
I'm much more sympathetic to that idea, but it's never yet come up in WG21 discussions. Until someone claims it's a real problem, I'd weakly prefer not to try to address it. I think addressing it will require either _completely_ uncoupling `SourceType` and `DestType` as you suggest (which makes the API seem to promise triviality in cases where the implementation can't actually guarantee it), or doing some sort of awful constrained template like `requires is_same_v<remove_cvref_t<SourceType>, remove_cvref_t<DestType>>` (which makes the API harder to understand).